src/locale/keymap-util.c

```diff
@@ -780,3 +782,228 @@ int x11_convert_to_vconsole(Context *c) {
+                if (r == 0)
+                        break;
+
+                line = strstrip(line);
+                if (strcasecmp_ptr(line, locale_entry) == 0)
+                        return 1;
+        }
+
+        /* if we are here, the locale is unsupported or invalid */
+        return -EINVAL;
+}
+#endif
+
+int locale_gen_enable_locale(const char *locale) {
+#if HAVE_LOCALEGEN
+        _cleanup_fclose_ FILE *fr = NULL, *fw = NULL;
+        _cleanup_(unlink_and_freep) char *temp_path = NULL;
+        _cleanup_free_ char *locale_entry = NULL;
+        bool locale_enabled = false, first_line = false;
+        struct stat st;
+        int r;
+
+        if (isempty(locale))
+                return 0;
+
+        if (locale_encoding_is_utf8_or_unspecified(locale)) {
+                locale_entry = strjoin(locale, " UTF-8");
+                if (!locale_entry)
+                        return -ENOMEM;
+        } else
+                return -ENOEXEC; /* We do not process non-UTF-8 locale */
+
+        r = locale_gen_entry_check_supported(locale_entry);
+        if ((r <= 0) && (r != -EOPNOTSUPP))
+                return r;
+
+        fr = fopen("/etc/locale.gen", "re");
+        if (!fr)
+                return -errno;
+
+        r = fopen_temporary("/etc/locale.gen", &fw, &temp_path);
+        if (r < 0)
+                return r;
+
+        /* apply mode of the original file to new file */
+        if (fstat(fileno(fr), &st) < 0)
+                return -errno;
+        (void) fchmod(fileno(fw), st.st_mode);
+
+        /* The config file ends with a line break, which we do not want to include before potentially appending a new locale
+         * instead of uncommenting an existing line. By prepending linebreaks, we can avoid buffering this file but can still write
+         * a nice config file without empty lines */
+        first_line = true;
+        for (;;) {
+                _cleanup_free_ char *line = NULL;
+                char *line_locale;
+
+                r = read_line(fr, LONG_LINE_MAX, &line);
+                if (r < 0)
+                        return r;
+                if (r == 0)
+                        break;
+
+                if (locale_enabled) {
+                        /* Just complete writing the file if the new locale was already enabled */
+                        if (!first_line)
+                                fputc('\n', fw);
+                        fputs(line, fw);
+                        first_line = false;
+                        continue;
+                }
+
+                line = strstrip(line);
+                if (isempty(line)) {
+                        fputc('\n', fw);
+                        first_line = false;
+                        continue;
+                }
+
+                line_locale = line;
+                if (line_locale[0] == '#')
+                        line_locale = strstrip(line_locale + 1);
+                else if (strcasecmp_ptr(line_locale, locale_entry) == 0)
```

Comment:
Please add (maybe as a separate commit):
```C
static inline bool strcaseeq_ptr(const char *a, const char *b) {
        return strcasecmp_ptr(a, b) == 0;
}
```
We don't use `str*cmp` for equality comparisons...